### PR TITLE
if-else rules for datasets with heterogeneous features

### PR DESCRIPTION
**Pull request by cgarcia-UCO**
# CONTEXT:

I am interested in if-else rules for classification and grammatical evolution. That's covered by PonyGE2. However, I am also interested in condtions on categorical features, such as x[:,2] == 'red', x[:,4] != 'tall', x[:,10] in {'cicle', 'triangle'}, or x[:,9] not in {'square', 'star'}. I am not aware of support for this in PonyGE2.

# ACHIEVEMENTS:

I have extended PonyGE2, which excellently manages grammars, with the ability of producing those conditions, along with '<=' and '>' for numerical features. To do that, I have taken inspiration on the tags GE_RANGE, and added the tags GE_GENERATE, which allows the grammar to be on-the-fly extended with such conditions, according to the values in the training dataset (conditions on the i-th feature considers the different values in the i-th column of the dataset; I have to say that this was a bit tricky to me, since a non-affix grammar approximation is not possible).

In a bit more detail, read_bnf_file method now reads the file into an StringIO buffer, which can be extended during the reading. There are functions which manages each of the previous types of conditions, all of them used by the new method try_processing_generate_cases, within the read_bnf_stringio method.

In addition, I have included some support for pandas DataFrames in get_data.py

Obviously, I have perform some test on a dataset of mine, and it worked. A grammar example can be found in if_else_classifier_heterogeneous_data.bnf, and the phenotype of the result of one test is (real labels have been "hidden", but notice the '==', '!=', 'np.isin', '~np.isin', '<=', and '>' generated conditions):

np.where((x.iloc[:,30] == 'str1') & (x.iloc[:,22] == 'str2') & (x.iloc[:,49] == 'str1') & (x.iloc[:,8] != 'str2'), np.where((~np.isin(x.iloc[:,30], ('str34', 'str33'))) & (x.iloc[:,85] == 'str34') & (x.iloc[:,22] == 'str2') & (x.iloc[:,59] != 'str3') & (x.iloc[:,75] <= 6.5) & (np.isin(x.iloc[:,6], ('str18', 'str4'))) & (np.isin(x.iloc[:,64], ('str34', 'str33'))), np.where((x.iloc[:,84] <= 6.0) & (x.iloc[:,27] == 'str31') & (x.iloc[:,64] == 'str34'), np.where((x.iloc[:,79] > 0.7), np.where((x.iloc[:,59] == 'str1'), 'str34', 'str34'), 'str34'), np.where((np.isin(x.iloc[:,20], ('str34', 'str33'))) & (x.iloc[:,40] <= -1000.0), np.where((x.iloc[:,59] == 'str26'), 'str34', 'str34'), np.where((np.isin(x.iloc[:,24], ('str34', 'str33'))), 'str33', 'str33'))), np.where((x.iloc[:,43] > 4.0) & (np.isin(x.iloc[:,14], ('str5', 'str1', 'str32'))) & (np.isin(x.iloc[:,8], ('str12', 'str13', 'str14', 'str15', 'str16'))) & (x.iloc[:,39] > 3.4), np.where((x.iloc[:,29] != 'str34') & (x.iloc[:,6] == 'str17'), np.where((np.isin(x.iloc[:,52], ('str34', 'str1'))), 'str34', 'str33'), np.where((x.iloc[:,18] != 'str19'), 'str34', 'str34')), np.where((x.iloc[:,48] <= 58.0) & (x.iloc[:,63] != 'str9'), np.where((x.iloc[:,74] > 26.5), 'str33', 'str33'), np.where((np.isin(x.iloc[:,69], ('str20', 'str21'))), 'str33', 'str33')))), np.where((~np.isin(x.iloc[:,55], ('str22', 'str23', 'str10', 'str34', 'str33', 'str6'))) & (x.iloc[:,69] == 'str34') & (x.iloc[:,42] > 50.6) & (np.isin(x.iloc[:,25], ('str7', 'str1'))) & (x.iloc[:,23] != 'str1') & (x.iloc[:,46] <= 24.0) & (x.iloc[:,5] == 'str8'), np.where((x.iloc[:,74] <= 2.9) & (np.isin(x.iloc[:,26], ('str24', 'str25'))) & (x.iloc[:,37] > 58.0), 'str34', 'str33'), np.where((x.iloc[:,73] == 'str29'), np.where((x.iloc[:,57] == 'str27'), np.where((~np.isin(x.iloc[:,23], ('str1', 'str33'))), 'str34', 'str34'), np.where((x.iloc[:,31] != 'str30'), 'str34', 'str33')), np.where((~np.isin(x.iloc[:,22], ('str1', 'str28'))), np.where((x.iloc[:,15] == 'str11'), 'str34', 'str34'), np.where((x.iloc[:,44] > 1.0), 'str33', 'str34')))))

I hope you like it.
Best regards.
Carlos.

### CODE REVIEW

**Comment by cgarcia-UCO** [comment]
I have just found that GE_GENERATE tags, which I introduced, do not work properly with ndarray datasets, nor datasets without categorical features.

I am fixing that.

**Comment by jmmcd** [comment]
Thanks for this great contribution! I was hoping to discuss it a bit before possibly pulling the code in. The overall approach of making a proper hook for generating parts of the grammar is a good one. I have two main queries:

First, is GENERATE a good name for this? The new functionality is quite specific to rules on categorical features, so probably it should be reflected in the name.

Second, could the new functionality actually be more general? Maybe someone else would want to use categorical features in a slightly different way. Can we brain-storm this and think about how we could enable more general use?

**Comment by jmmcd** [comment]
Continuing with some more thoughts...

3. There is a lot of copy-pasted code. I'm sure a shorter implementation would be possible.

4. This is not integrated with the existing system GE_RANGE (or vice versa), instead it's in addition to it. 

In both of these points, my concern is to avoid adding a lot of code and an increase in complexity, if not needed.

**Comment by jmmcd** [comment]
Idea: what if we allow user code in the grammar?

Simple example first to implement the functionality of `GE_RANGE`:

```
<f> ::= <a> | b
<a> ::= c | d | GE_GENERATE [f"x[:, {idx}]" for idx in range(dataset_n_vars)]
```

In `grammar.py`, when we encounter `GE_GENERATE` we `eval` the code that follows, which must produce a list, and we take each element of that list as a new terminal. This *replaces* the current `GE_RANGE` code.

Complex example to implement the categorical variables:

```
<f> ::= <a> | b
<a> ::= c | d | GE_GENERATE [f"x[:, {idx}] == {val}" for idx in range(dataset_n_vars) for val in dataset_x[idx]]
```

So, now we would need `dataset_x` to be a dict holding all the possible values for each variable, eg `dataset[x]` might be `['red', 'blue', 'green']`. That would have to be created in the same way that `n_vars` currently is. And of course we would need to write a few more expressions similar to the above, to handle the not-equals case, the `isin` case, and so on.

Would this work?

Advantage: it would result in a lot less new code, and it would be much more generic. It would unify `GE_RANGE` and `GE_GENERATE`. A user who wants something new can get it without writing a patch, just by writing in the grammar.

(For backwards compatibility, we would have to keep `GE_RANGE` as a special case.)

**Comment by cgarcia-UCO** [comment]
A very enriching discussion! Thank you!

I think that your idea of allowing user code in the grammar is very elegant, and it obviously reduces the amount of new code. I can come up with three potential issues (but I really think that pros are larger):
* Initially, the grammar would have to take notice (even deeper) of how the data should be accessed. In particular, `x[:, {idx}]` is for addressing a numpy ndarray, but `x.iloc[:,{idx}]` should be used for pandas DataFrames. I say initially because the library (the evaluate function) could provide an abstraction level, which hid these details.
* `in` and `not in` conditions would  require either (as I understand):
  * a complex notation in the grammar: something involving code such as that in the `get_all_groups` function (the name might not be the best one): 

     `chain.from_iterable(combinations(collection, r) for r in range(min_size,max_size))`

  * an abstraction such as the own `get_all_groups` function
  * a grammar tag to generate those subset
* We should distinguish between categorical and numeric features. To my opinion, equality (inequality, `in`, and `notin`) conditions are not interesting for numeric features. Therefore, I think that some other structures should be created, in the same way that `n_vars` is currently and `dataset_x` would be, to be able to iterate just over categorical features or just over numeric ones.

**Comment by jmmcd** [comment]
Thanks again.

Looking at the whole PR, there are a few different issues to discuss, though they are intertwined: Pandas DataFrames, ARFF, generating grammar items for categorical features, generating `<=` and `>` grammar items for numerical features, and generating grammar items for heterogeneous datasets.

ARFF would be nice to have and seems a simple addition. Does ARFF necessarily require DataFrames? I think you are introducing the DataFrame only in the case of ARFF, right? And you have a comment "Arff support just for train set." Could you explain the reasoning here? 

PonyGE currently has its own way of doing data, ie it assumes separate train and test files specified by a command-line parameter. But we will move to a more sklearn-like style, where reading data can be done independently by the user and then `.fit(X, y)`. Sklearn and matplotlib have moved to be able to accept either Numpy arrays or DataFrames here, and I think it usually involves lots of if-else checks behind the scenes. So probably we will follow this.


About `get_all_groups`, it would certainly be good to add a helper function like this if needed. We would import it in grammar.py so it's available during the `eval`. However we shouldn't have any tricks like reducing the number of items per group as this introduces silent changes which the user wouldn't expect.

I also would not be in favour of any heuristic like that used for categorical features, ie less than 10 distinct values implies a categorical feature. It will have to be explicit, somehow. One advantage of doing all this with user code in the grammar is that it is explicit and is done by the user, no behind-the-scenes guessing. So, eg if we read in the dataset to a DataFrame and we store the types as `dataset_types` the user could write:

```
<f> ::= <a> | <b>
<a> ::= c | d | GE_GENERATE [f"x[:, {idx}] == {val}" for idx in range(dataset_n_vars) for val in dataset_x[idx] if dataset_types[idx] == str]
<b> ::= GE_GENERATE [f"x[:, {idx}" for idx in range(dataset_n_vars) if dataset_types[idx] == float]
```

Again this would reduce a lot of complexity (I hope), I mean all the code about 

For the `<=` and `>` comparisons,

**Comment by jmmcd** [comment]
Some of these pieces of user code in the grammar are becoming complex! So I'm tempted to meet the original PR halfway, by introducing a few "shortcuts", eg

<comparison> ::= GE_GENERATE_COMPARISONS

In `grammar.py` if we meet this one, then we execute the code `<comparison> ::= [f"x[:, {idx}] <= {val}" for idx in range(dataset_n_vars) for val in set(dataset_x[idx]) if dataset_types[idx] == float]`

**Comment by jmmcd** [comment]
Noting a few concrete Pandas implementations:

```python
import pandas as pd
from itertools import chain, combinations

x = pd.DataFrame({"a": ["x", "y", "z"], "b": [2.3, 1.1, 5.6], "c": [7, 8, 3]})
print(x)

# vars
VARS = [f"x['{idx}']" for idx in x.columns]
print(VARS)

# var <= val for numerical vars only
LT = [f"x['{idx}'] <= {val}" for idx in x.columns
      for val in sorted(x[idx].unique())
      if x.dtypes[idx] != object]
print(LT)
print(eval(LT[0]))

# var = val for categorical vars only
EQ = [f"x['{idx}'] == '{val}'" for idx in x.columns
      for val in x[idx].unique()
      if x.dtypes[idx] == object]
print(EQ)
print(eval(EQ[0]))
    
# from itertools but we add minlen and maxlen
def powerset(iterable, minlen=0, maxlen=None):
    "powerset([1,2,3]) --> () (1,) (2,) (3,) (1,2) (1,3) (2,3) (1,2,3)"
    s = list(iterable)
    if maxlen is None: maxlen = len(s)
    return chain.from_iterable(combinations(s, r) for r in range(minlen, maxlen+1))

IN = [f"x['{idx}'].isin({s})" for idx in x.columns
      for s in powerset(x[idx].unique(), minlen=2) # no point in sets of len 0 or 1
      if x.dtypes[idx] == object]
print(IN)
print(eval(IN[0]))
```

(I think powerset is the right name for this, better than `get_all_groups`.)

**Comment by jmmcd** [comment]
After generating many productions, we should check that max codon size is large enough. It should either be *equal* to the number of productions in the largest rule, or an integer multiple, or a large (non-integer) multiple. If it's *smaller* some productions are impossible, and if it's a small non-integer multiple (eg 1.2x the number of productions) that is an important bias.

**Comment by jmmcd** [comment]
> For the `<=` and `>` comparisons, I don't understand how we choose the right-hand side values. I think they may be chosen from the set of values of that feature in the dataset?

```python
LT = [f"x['{idx}'] <= {val}" for idx in x.columns
      for val in sorted(x[idx].unique())
      if x.dtypes[idx] != object]
```

For a float variable in a large dataset, this could create many thousands or millions of productions!

**Comment by cgarcia-UCO** [comment]
Thanks to you. I really enjoy participating in the development of PonyGE2.

So, independent comments:
* Yes. I introduced ARFF reading and DataFrame because of requirements of another of my projects. But they are not dependent. My ARFF support does read data into a DataFrame (it did not have to necessarily), but DataFrame support in `grammar.py` and `supervised_learning.py` is independent of having had read the dataset from an ARFF file.

* Regarding the comment "Arff support just for train set", my fault. Testing set is being read (if not None) from an ARFF file. I just forgot to remove that comment. In any case, I am assuming both sets come in the same file format.

* Great for moving to being able to accept either Numpy arrays or DataFrames. I do not know if working on Numpy arrays would be more efficient than working on DataFrames. I think so. In any case, I think that always working on DataFrames would be more efficient than lots of if-else checks behind the scenes. And yes, working with DataFrames allows me to address my goals more easily.

* I accept that heuristics such as "less than 10 distinct values makes a feature categorical" is not right way. In my project, this is actually solved by reading an ARFF file into a DataFrame. Your solution with `dataset_types` is ok to me. Another possibility is that the user specifies the index of the numerical features with another production rule:
     
    ```
    <idx_numerical_features> ::= 3 | 4 | 7 | 8
    <idx_categorical_features> ::= 1 | 2 | 5 | 6
    ```  
   In any case, this is actually not needed when reading ARFF files into DataFrames.

* For the `<=` and `>` comparisons, yes, I am currently considering every distinct value in the dataset, which can create thousands or millions of productions. This is not currently a problem in my other project (except for the max codon size comment), but would be in others. I think that a better general approach would involve generating numbers within a given/induced range, as you suggest above. In any case, to me, this would still be an open issue.

* Regarding the tricks in `get_all_groups`, the problem is that generating the powerset of a not small set may consume the whole memory (even swap) of common computers (and freeze the computer). In fact, it happened to me with sets of 20 distinct values. So I really think that some protection mechanism should be considered. Then, perhaps a better name, even better than `powerset`, would be `get_subsets(set, max_size)` without a default `max_size` value and with a warning.

* Regarding half meeting the original PR, I really like allowing the user to write some code in the BNF grammar file, although it is true that it makes writing new grammars more complex. Perhaps a good solution is to introduce both, PonyGE2-automatically-processing tags and allowing the user to write code.
  * By the way, I am interesting in introducing conditions such as `x[:,13] < x[:,14]`, which would be easy if you know apriori the indexes of numerical features (they do even not require code in the grammar), and it is not currently supported in the PR.
  
* Regarding the max codon size, I had not noticed that. In any case, I think that that can easily be fixed, due to `grammar.codon_size` is not used before reading the BNF grammar, is it?

And that's all for now. I would happily participate in developing a new PR with these ideas (perhaps I actually do that, since I actually enjoy doing that), but I think I would not be able before new year. Meanwhile, and as I said, I really appreciate participating in this discussion.

**Comment by jmmcd** [comment]
Ok! 

* We will definitely add ARFF. If I can take it from the existing PR I will, else we'll make a new small patch just for this. But let's wait until the DataFrame thing is decided.
* Using always DataFrame behind the scenes does seem best/easiest, though I suppose there could be users who could be disrupted by it. (Comments welcome from all.) It would mean changing our grammars *back* to the `x[0]` style instead of `x[:, 0]`. (We went in the opposite direction a few weeks ago.) 
* `<idx_numerical_features> ::= 3 | 4 | 7 | 8`: of course users can always write this to create their expressions manually. But I wouldn't like to have code in GE_GENERATE which refers to the content of other rules, it would become quite hairy.
* "generating numbers within a given/induced range, as you suggest above" - let's go with this.
* For `powerset` or `get_subsets`, yes, I would prefer to have an exception rather than a silent change.
* `conditions such as x[:,13] < x[:,14]`: we can certainly generate these using a similar comprehension to that suggested above for the other `<=` conditions.
* About max codon size, we can just add a check at the end of grammar parsing for the number of productions in each rule. We should have this even if we don't have generated productions.

I'm happy to wait a bit especially on the DataFrame issue as that would be potentially disruptive.

**Comment by cgarcia-UCO** [comment]
Hello again.

According to the previous comments, I understand that this commit is not going to be accepted, which I think is the right decission. I am not sure if it should be closed and move the discussion to another GitHub section, perhaps issues.

In any case, I am closing the PR because I mistakenly added some changes in my master branch due to an unwanted branch merge.

Best regards.
